files/opt/dash-button/scripts/script.js

```diff
@@ -0,0 +1,67 @@
+const button = require('node-dash-button');
+const https = require('https');
+const request = require('request');
+var fs = require('node-fs');
+
+// Load config.json
+const configJSON = fs.readFileSync(process.cwd() +"/scripts/config.json");
+const config = JSON.parse(configJSON);
+const buttonConfig = config.buttons;
+const providerConfig = config.providerConfig;
+
+// Add handler for dash-buttons
+buttonConfig.forEach(function(value) {
+    console.log('Configuring button {' + value.mac + '} for product {' + value.providerConfig.productName + '} with provider ' + value.provider);
+    var dash = button(value.mac, null, null, 'all');
+    dash.on("detected", function (){
+        dashHandler(value);
+    });
+});
+
+// Switch handler for dash-buttons
+function dashHandler(sender) {
+    switch(sender.provider) {
+        case "bring":
+            addToBringList(sender);
+            break;
+    }
+}
```

Comment:
Das finde ich so noch nicht gut. Ich würde folgendes vorschlagen:

1. Jeder Provider wird in einer eigenen Datei `providers/[name].js` abgelegt.
2. Beim Start werden alle diese Dateien geladen.

Die Implementierung könnte dann so aussehen…

**script.js**

```js
let providers = { };
// Hier alle Provider-JS-Dateien laden
```

Wenn dann ein Button geklickt wird…

```js
providers[sender.provider](sender);
```

Frage hier: Enthält das Objekt `sender` auch schon gleich die gesamte Konfiguration des Buttons? D.h. auch die Provider-Konfiguration, die nicht spezifisch für diesen Button ist, sondern für alle gilt?

Ziel: Die *Provider-Funktion* sollte nur mit den Werten arbeiten, die ihr als Paramter übermittelt werden und nicht auf irgenwelche globalen `config`-Objekte zugreifen.

**providers/bring.js**

```
providers.bring = function (sender) {
  // Request abschicken...

  function sendNotification(sender) {
    // ...
  }

};
```

So ist sichergestellt, dass jeder Provider unabhängig implementiert werden kann und keine Änderung von `script.js` notwendig ist um einen zusätzlichen Provider zu implementieren. Zudem können sie sich nicht in die Quere kommen, da 1. der gesamte Code eines Providers in einer Funktion gekapselt ist und 2. nicht auf globale Objekte zugegriffen wird.